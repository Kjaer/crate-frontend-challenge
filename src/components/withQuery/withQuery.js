import React, { Suspense } from 'react';
import useFetch from 'fetch-suspense';
import { API } from '../../utils/constants';

import 'chota/src/_util.css';
import hourglass from './hourglass.gif';

function Loading() {
  return(
    <section className="text-center" data-test="crate-fetch-loading">
      <h4>LOADING ... </h4>
      <img src={hourglass} alt="data is loading from remote..."/>
    </section>
  )
}

function executeQuery(Component) {
  return function (props) {
    const response = useFetch(API, {
      method: 'POST',
      body: JSON.stringify({stmt: props.query})
    });
    
    return(
      <Component data={response} {...props} />
    )
  }
}


export function withQuery(Component) {
  const QueryComponent = executeQuery(Component);

  return function (props) {
    return(
      <Suspense fallback={<Loading/>}>
        <QueryComponent {...props}/>
      </Suspense>
    )
  }
}

export default withQuery;
