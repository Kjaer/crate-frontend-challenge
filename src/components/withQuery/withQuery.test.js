import React from 'react';

import { configure, render, wait } from '@testing-library/react';
import QueryResponseBuilder from "../../tests/queryBuilder";

import withQuery from "./withQuery";

function Dummy({data}) {
  const { rows: [ [ { name } ] ] } = data;

  return(
    <p data-test="crate-dummy">
      {name}
    </p>
  )
}

const QueryWrapper = withQuery(Dummy);

describe('withQuery', function () {

  beforeAll(() => {
    configure({testIdAttribute: 'data-test'});
  });

  afterAll(()=>{
    fetch.resetMocks()
  });


  it('pass fetched data to the wrapped component', async () => {
    const payload = new QueryResponseBuilder().withRows({name: 'dummy content'});
    fetch.mockResponseOnce(JSON.stringify(payload), {headers: {'Content-Type': 'application/json'} });

    const { getByText, container } = render(<QueryWrapper query="SELECT * FROM DUMMY"/>);
    // testing the loading
    expect(container).toMatchSnapshot();

    await wait(() => {
      expect(getByText(/dummy content/i)).toBeInTheDocument();
      // testing the wrapped component
      expect(container).toMatchSnapshot();
    });

  })
});
