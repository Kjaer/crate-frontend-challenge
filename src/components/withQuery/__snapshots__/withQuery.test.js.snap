// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withQuery pass fetched data to the wrapped component 1`] = `
<div>
  <section
    class="text-center"
    data-test="crate-fetch-loading"
  >
    <h4>
      LOADING ... 
    </h4>
    <img
      alt="data is loading from remote..."
      src="hourglass.gif"
    />
  </section>
</div>
`;

exports[`withQuery pass fetched data to the wrapped component 2`] = `
<div>
  <p
    data-test="crate-dummy"
  >
    dummy content
  </p>
</div>
`;
